### PR TITLE
[WIP]: Integration test

### PR DESCRIPTION
**Pull request by MartinaFelberFEM**
Working on a simple integration test:
- Implemented a client which connects to a server and prints it's full address space (at the moment only from ObjectsFolder node and only the nodeId - but this will be extended)
- Built a reference server with compiled nodesets (DI and PLCopen at the moment)
- Device under test is a server with NodesetLoader plugin, which imports the same nodesets as the reference server

Test case:
- start reference server, connect with client and print it's addressspace to a file
- start test server, connect with client and print it's address space to a file
- compare both files

Questions:
- Is the test directory location ok? I have put the sources at /backends/open62541/tests/integration. I am not sure if this is a good location. Because I think it's not a unit test, but an integration test
- Test automation? Do you have a suggestion how I should start/stop server/client etc? from CMakeLists file? Execute a shell script? Or can I execute everything with CMake? Or shall I use the unit testing framework?

Thanks for your help

PS: src and CMakeLists files are work in progress

### CODE REVIEW

**Comment by MartinaFelberFEM** [comment]
I would have another question regarding coding style. 
Do you comply to a specific coding style? I think I have accidentally used a different style for the curly braces at the beginning of the function body. I have used an auto formatter. I can change that.

**Comment by matkonnerth** [comment]
Cool, I like the idea of testing the nodeset import this way. The location is ok.

Regarding the test execution:
This should be possible with a shell script, shell scripts can then also be invoked from cmake. I'm not the shell guru, but we will need something like:
- start the reference server
- start the test client, client dumps the addresspace
- stop(kill) the reference server, stop client
- start server with nodesetimport
- start client, dump addressspace
- stop server2, stop client
- do comparison (could be a cmake test)

I think this is the right cmake command for executing a shell script: https://cmake.org/cmake/help/latest/command/add_custom_command.html

I think the reference address space shouldn't change that often, therefore we could also version the file of the reference address space. Then we wouldn't need the whole address space generation with the open tools all the time, also if we think of to remove the open repository from this repo.

**Comment by matkonnerth** [comment]
> I would have another question regarding coding style.
> Do you comply to a specific coding style? I think I have accidentally used a different style for the curly braces at the beginning of the function body. I have used an auto formatter. I can change that.

There's a .clang-format file in the root directory. It depends on your ide how to use, I use vscode and there's a clang-format extension for vscode. Hope this helps.

**Comment by MartinaFelberFEM** [comment]
Many thanks for your help!

I am using a shell script. I hope the linux OS dependency isn't a problem.
Maybe I refactor it later ...

I am using vs studio code with the formatter config now. Works great.

File: backends/open62541/tests/integration/client/browse_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
+#include "browse_utils.h"
+#include "sort_utils.h"
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <open62541/client_highlevel.h>
+#include <open62541/types.h>
+
+using namespace std;
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/browse_utils.cpp:9]: don't import the whole namespace std (I know it's only in the impl file, but I think we should avoid this completely). You can use using _std::cout_, _std::vector_ instead or simply type the std:: in front of.

File: backends/open62541/tests/integration/client/browse_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    UA_NodeId_copy(&Id, &(bReq.nodesToBrowse[0].nodeId));
+    bReq.nodesToBrowse[0].browseDirection = UA_BROWSEDIRECTION_FORWARD;
+    bReq.nodesToBrowse[0].includeSubtypes = UA_TRUE;
+    bReq.nodesToBrowse[0].referenceTypeId =
+        UA_NODEID_NUMERIC(0, UA_NS0ID_REFERENCES);
+    bReq.nodesToBrowse[0].resultMask = UA_BROWSERESULTMASK_ALL;
+    UA_BrowseResponse bResp = UA_Client_Service_browse(pClient, bReq);
+    if (bResp.responseHeader.serviceResult == UA_STATUSCODE_GOOD)
+    {
+        TReference CopyReference; // copy browse results to output vector
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/browse_utils.cpp:51]: I think TReference could hold the UA_NodeIds, not UA_NodeId*

File: backends/open62541/tests/integration/client/browse_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        for (size_t i = 0; i < bResp.resultsSize; i++)
+        {
+            for (size_t refIdx = 0; refIdx < bResp.results[i].referencesSize;
+                 refIdx++)
+            {
+                UA_ReferenceDescription *ref =
+                    &bResp.results[i].references[refIdx];
+
+                CopyReference.pReferenceTypeId = UA_NodeId_new();
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/browse_utils.cpp:60]: then this is not needed

File: backends/open62541/tests/integration/client/browse_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        ret = UA_FALSE;
+    }
+
+    UA_BrowseRequest_deleteMembers(&bReq);
+    UA_BrowseResponse_deleteMembers(&bResp);
+
+    return ret;
+}
+
+void FreeReferencesVec(TReferenceVec &References)
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/browse_utils.cpp:92]: and this is also not needed

File: backends/open62541/tests/integration/client/browse_utils.h [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#ifndef _BROWSE_UTILS_H
+#define _BROWSE_UTILS_H
+
+#include <open62541/client_config.h>
+#include <open62541/types.h>
+#include <vector>
+
+typedef struct
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/browse_utils.h:8]: could be simple a struct Reference { UA_NodeId type, UA_NodeId target}; I know I've use this TNodeId style in the nodesetLoader, but I think the T is pretty oldschool and not needed.

File: backends/open62541/tests/integration/client/browse_utils.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+typedef std::vector<TReference> TReferenceVec;
+
+UA_Boolean BrowseReferences(UA_Client *pClient, const UA_NodeId &Id,
+                            TReferenceVec &oReferences);
+
+void FreeReferencesVec(TReferenceVec &References);
+
+UA_Boolean IsSubType(const UA_NodeId &BaseType, const UA_NodeId &SubType);
+
+#endif // _BROWSE_UTILS_H
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/browse_utils.h:23]: clang build is happy when there's new line at the eof

File: backends/open62541/tests/integration/client/client.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    {
+        cout << "Error: connection could not be established" << endl;
+        UA_NodeId_clear(&StartId);
+        UA_Client_delete(client);
+        return EXIT_FAILURE;
+    }
+
+    int ret = EXIT_SUCCESS;
+    TNodeIdContainer NodeIdContainer;
+    NodeIdContainer.clear();
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/client.cpp:194]: not necessary

File: backends/open62541/tests/integration/client/operator_ov.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    default:
+        os << "Error";
+    }
+    return os;
+}
+
+ostream &operator<<(ostream &os, const UA_QualifiedName &QualifiedName)
+{
+    os << "ns=" << QualifiedName.namespaceIndex << ";"
+       << string((char *)QualifiedName.name.data, QualifiedName.name.length);
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/operator_ov.cpp:67]: there's an operator overload for << UA_String

File: backends/open62541/tests/integration/client/sort_utils.cpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+#include <cassert>
+#include <iostream>
+
+using namespace std;
+
+/*****************************************************************************/
+// return 0 if both identifier types are equal
+// return 1 if lhs identifier type shall be inserted before rhs
+// return -1 if lhs identifier type shall be inserted after rhs
+int SortIdentifierType(const UA_NodeIdType lhs, const UA_NodeIdType rhs)
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/sort_utils.cpp:12]: I think there are UA_NodeId_order, can this be used instead?

File: backends/open62541/tests/integration/client/sort_utils.h [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#ifndef _SORT_UTILS_H
+#define _SORT_UTILS_H
+
+#include <open62541/types.h>
+#include <vector>
+
+typedef std::vector<UA_NodeId *> TNodeIdContainer;
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/client/sort_utils.h:7]: prefer _using TNodeIdContainer = std::vector<UA_NodeId*>_ over the typedef way

File: backends/open62541/tests/integration/reference_server/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    INTERNAL
+)
+
+add_executable(referenceServer server.cpp
+    ${UA_NODESET_INTEGRATION_TEST_DI_SOURCES}  ${UA_TYPES_INTEGRATION_TEST_DI_SOURCES}
+    ${UA_NODESET_INTEGRATION_TEST_PLC_SOURCES} ${UA_TYPES_INTEGRATION_TEST_PLC_SOURCES})
+target_link_libraries(referenceServer PRIVATE open62541::open62541)
+target_include_directories(referenceServer PRIVATE ${CMAKE_BINARY_DIR}/src_generated)
+
+#add_dependencies(referenceServer open62541-generator-integration-ns)
```
Comment (matkonnerth) [inline backends/open62541/tests/integration/reference_server/CMakeLists.txt:32]: https://cmake.org/cmake/help/latest/command/add_dependencies.html

**Comment by matkonnerth** [comment]
Please be aware that I've added a cmake option to enable the the integration tests. There is also a new build that runs the integration tests on travis.

option(ENABLE_INTEGRATION_TEST off "run detailled tests to compare address spaces")

**Comment by matkonnerth** [comment]
Thank you for this first initial commit to improve the quality of the nodesetLoader, great job!

**Pull request merged**